// listhen/dist/cli.cjs
'use strict';

const citty = require('citty');
const pathe = require('pathe');
const index = require('./index.cjs');
require('node:http');
require('node:https');
require('node:util');
require('get-port-please');
require('http-shutdown');
require('defu');
require('consola/utils');
require('node:child_process');
require('node:fs');
require('node:os');
require('node:path');
require('node-forge');
require('consola');
require('node:fs/promises');

const name = "listhen";
const version = "1.2.1";
const description = "👂 Elegant HTTP Listener";

const main = citty.defineCommand({
  meta: {
    name,
    description,
    version
  },
  args: {
    cwd: {
      type: "string",
      description: "Current working directory"
    },
    entry: {
      type: "positional",
      description: "Listener entry file (./app.ts)",
      required: true
    },
    port: {
      type: "string",
      description: "Port to listen on (use PORT environment variable to override)"
    },
    host: {
      type: "string",
      description: "Host to listen on (use HOST environment variable to override)"
    },
    clipboard: {
      type: "boolean",
      description: "Copy the URL to the clipboard",
      default: false
    },
    open: {
      type: "boolean",
      description: "Open the URL in the browser",
      default: false
    },
    baseURL: {
      type: "string",
      description: "Base URL to use"
    },
    name: {
      type: "string",
      description: "Name to use in the banner"
    },
    https: {
      type: "boolean",
      description: "Enable HTTPS",
      default: false
    },
    watch: {
      type: "boolean",
      description: "Watch for changes",
      alias: "w",
      default: false
    }
  },
  async run({ args }) {
    const opts = {
      ...args,
      port: args.port,
      hostname: args.host,
      clipboard: args.clipboard,
      open: args.open,
      baseURL: args.baseURL,
      name: args.name,
      https: args.https
      // TODO: Support custom cert
    };
    const entry = pathe.isAbsolute(args.entry) || args.entry.startsWith(".") ? args.entry : `./${args.entry}`;
    if (args.watch) {
      await index.listenAndWatch(entry, opts);
    } else {
      const devServer = await index.createDevServer(entry, opts);
      await index.listen(devServer.nodeListener, opts);
      await devServer.reload(true);
    }
  }
});
const runMain = () => citty.runMain(main);

exports.main = main;
exports.runMain = runMain;
